### PR TITLE
Support specifying per-frame positional arguments in sequence processing test utility

### PR DESCRIPTION
**Pull request by stiepan**
Signed-off-by: Kamil Tokarski <ktokarski@nvidia.com>

## Category:

**Other** (*e.g. Documentation, Tests, Configuration*)

## Description:

Refactors the way to specify arguments in sequence processing tests utility to enable testing positional arguments that may (or may not) be on GPU.
Moreover, the callback that produces arguments now receives extra context with the index of batch, sample and frame (if any).

## Additional information:
Required for testing warp positional matrix argument https://github.com/NVIDIA/DALI/pull/3879

### Affected modules and functionalities:
Sequence test utility along with existing test suites for Gaussian blur, Laplacian and rotate.

### Tests
- [ ] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [x] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2786

### CODE REVIEW

**Comment by stiepan** [comment]
!build

**Review by banasraf — APPROVED** [review]

File: dali/test/python/sequences_test_utils.py [hunk 1]
```diff
@@ -28,13 +29,55 @@
...
+        self.desc = arg_desc
+        self.data = data
+
+
 class ParamsProviderBase:
     """
     Computes data to be passed as argument inputs in sequence processing tests, the `compute_params` params
-    should return tuple of lists with data for respectively per-sample and per-frame arguments.
-    The `expand_params` should return corressponding unfolded/expanded arguments to be used in the
+    should return a lists of ArgData; the `arg_desc.data` dimensionality must match `arg_data.desc.is_per_frame`.
```
Comment (szalpal) [inline dali/test/python/sequences_test_utils.py:76]:
```suggestion
    should return a lists of ArgData; number of dimensions in the `arg_desc.data` and `arg_data.desc.is_per_frame` arguments must match.
```

File: dali/test/python/sequences_test_utils.py [hunk 2]
```diff
@@ -28,13 +29,55 @@
...
+    The `expand_params` should return corressponding unfolded/expanded ArgData to be used in the
```
Comment (szalpal) [inline dali/test/python/sequences_test_utils.py:77]:
```suggestion
    The `expand_params` should return corresponding unfolded/expanded ArgData to be used in the
```

File: dali/test/python/sequences_test_utils.py [hunk 3]
```diff
@@ -28,13 +29,55 @@
...
+        self.cb = cb
+
+    def __repr__(self):
+        return "ArgCb{}".format((self.cb, self.desc))
+
+
+class ArgData:
+    def __init__(self, arg_desc : ArgDesc, data):
+        self.desc = arg_desc
+        self.data = data
```
[the remark this answers is not in the document]
Comment (stiepan) [reply dali/test/python/sequences_test_utils.py:70]: Absolutely, thanks.

[the remark this answers is not in the document]
Comment (stiepan) [reply dali/test/python/sequences_test_utils.py]: done

File: dali/test/python/sequences_test_utils.py (hunk 2 above)
Comment (stiepan) [reply dali/test/python/sequences_test_utils.py]: done

File: dali/test/python/sequences_test_utils.py (hunk 1 above)
Comment (stiepan) [reply dali/test/python/sequences_test_utils.py]: changed it to
"should return a lists of ArgData; number of dimensions in the `arg_desc.data` must reflect `arg_data.desc.is_per_frame` argument."

Comment (stiepan) [reply dali/test/python/sequences_test_utils.py]: Because, in the hindsight maybe match is not the most fortunate word in this context.

**Comment by stiepan** [comment]
!build

**Review by szalpal — APPROVED** [review]

**Pull request merged**